### PR TITLE
Sdb update

### PR DESCRIPTION
**Pull request by radare**

### CODE REVIEW

File: shlr/sdb/src/ht.c [hunk 1]
```diff
@@ -1,239 +1,258 @@
...
+	SdbListIter* iter;
+	ut32 computed_hash = hash ? *hash : ht->hashfn (key);
+	ut32 bucket = computed_hash % ht->size;
+	SdbList* list = ht->table[bucket];
+	ls_foreach (list, iter, kvp) {
+		if (key == kvp->key || (ht->cmp && !ht->cmp (key, kvp->key))) {
+#define EXCHANGE 1
+#if EXCHANGE
+			ls_split_iter (list, iter);
+			ls_append (ht->deleted, iter);
```
Comment (ret2libc) [inline shlr/sdb/src/ht.c]: what is the purpose of `ht->deleted`?

File: shlr/sdb/src/ht.c [hunk 2]
```diff
@@ -1,239 +1,258 @@
...
-	if (ht) {
-		free (ht->table);
-		ls_free (ht->list);
-		free (ht);
+void ht_free(SdbHash* ht) {
+	ut32 i;
+	for (i = 0; i < ht->size; i++) {
+		ls_free (ht->table[i]);
 	}
+	ht_free_deleted (ht);
```
Comment (ret2libc) [inline shlr/sdb/src/ht.c]: i think this leaks memory, because `ht->deleted` is re-allocated in `ht_free_deleted`.

**Comment by radare** [comment]
i explained this to you yesterday :PP

> On 25 Oct 2016, at 10:57, Riccardo Schirone <notifications@github.com> wrote:
> 
> @ret2libc commented on this pull request.
> 
> In shlr/sdb/src/ht.c <https://github.com/radare/radare2/pull/6006#pullrequestreview-5606818>:
> 
> > -	ht->deleted_entries = 0;
> -	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
> -		if (entry_is_present (e))
> -			ht_insert (ht, e->hash, e->data, e->iter);
> +bool ht_delete_internal(SdbHash* ht, const char* key, ut32* hash) {
> +	SdbKv* kvp;
> +	SdbListIter* iter;
> +	ut32 computed_hash = hash ? *hash : ht->hashfn (key);
> +	ut32 bucket = computed_hash % ht->size;
> +	SdbList* list = ht->table[bucket];
> +	ls_foreach (list, iter, kvp) {
> +		if (key == kvp->key || (ht->cmp && !ht->cmp (key, kvp->key))) {
> +#define EXCHANGE 1
> +#if EXCHANGE
> +			ls_split_iter (list, iter);
> +			ls_append (ht->deleted, iter);
> what is the purpose of ht->deleted?
> 


store a list of deleted iters so the callers can iterate their lists without worrying about elements being deleted magically, fixing UAFs problems when deleting more than one key in the froeach callback. this is why there’s also a reference couting to define the depth and only free that list when it’s safe
> In shlr/sdb/src/ht.c <https://github.com/radare/radare2/pull/6006#pullrequestreview-5606818>:
> 
> >  	}
> +	ht_free_deleted (ht);
> i think this leaks memory, because ht->deleted is re-allocated in ht_free_deleted.
> 

agree. i also found another leak in the implementation. thanks

**Comment by radare** [comment]
both fixed in master

**Comment by ret2libc** [comment]
> store a list of deleted iters so the callers can iterate their lists without worrying about elements being deleted magically, fixing UAFs problems when deleting more than one key in the froeach callback. this is why there’s also a reference couting to define the depth and only free that list when it’s safe

I'm not sure it really makes sense... it's like saying "we never free memory to avoid UAF" :P (indeed ht->deleted is only freed at the very end of the liftime of an ht. if i add 30000 keys and then delete all of them, all the memory will still be there for nothing). Also, in `ht_delete_internal` you can remove just one element at a time, so you shouldn't have any problem.

**Comment by radare** [comment]
no. it means, we delegate the freeing of memory, what other solution do you propose? because this is the only way we can do to safely delete while iterating. and yes you can have problems if you remove the next 2 elements of the  list because the iters will be dangled. and removing a single element is also a problem for the old way.

we should delete all the keys in the deleted ht when depth == 0, which means its running outside an sdb_foreach, so it will not “leak” (aka delegate that much)

**Comment by ret2libc** [comment]
IMO the callback of sdb_foreach shouldn't modify the hashtable, because there are too many ways this may go wrong. If someone wants to delete things from the hashtable, it can use sdb_foreach to create a list `to_remove` and then, after having traversed all elements, he can iterate over `to_remove` to remove all elements that were stored there.

**Comment by radare** [comment]
then removing elements turns out to be very slow. in this way we cover most common use cases without losing on performance too much. Which case will not be covered?

File: shlr/sdb/src/ht.c (hunk 1 above)
Comment (radare) [inline shlr/sdb/src/ht.c]: to store a list of the deleted iters? isnt that obvious with the line itself?

**Comment by radare** [comment]
Delegating the effort to create a delete list to the user makes the library more complicated to use. I will prefer to try to make things as simpler as possible to the user

**Comment by radare** [comment]
MERGED

**Pull request closed without merging**